### PR TITLE
Resolve ILLink Warnings on Microsoft.Extensions.Http

### PR DESCRIPTION
**Pull request by joperezr**
Resolving ILLink warnings on Microsoft.Extensions.Http.

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.cs [hunk 1]
```diff
@@ -398,16 +403,22 @@ public static IHttpClientBuilder ConfigureHttpMessageHandlerBuilder(this IHttpCl
         {
             ReserveClient(builder, typeof(TClient), builder.Name, validateSingleType);
 
-            builder.Services.AddTransient<TClient>(s =>
+            builder.Services.AddTransient<TClient>(s => AddTransientHelper<TClient, TImplementation>(s));
+
+            return builder;
+
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2091:RequiresUnreferencedCode",
```
Comment (vitek-karas) [inline src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.cs:410]: I don't think it should use `RequiresUnreferencedCode` we typically use that only when it involves the RUC attribute. In this case this is typical dataflow warning (on generic parameters).

**Review by vitek-karas — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.cs (hunk 1 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.cs:410]:
```suggestion
            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2091:UnrecognizedReflectionPattern",
```

We use `UnrecognizedReflectionPattern` elsewhere:

https://github.com/dotnet/runtime/blob/d4fc640c49f8de360a4a220dbcf48203be6f42c4/src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs#L72

File: src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.cs [hunk 2]
```diff
@@ -335,16 +335,21 @@ public static IHttpClientBuilder ConfigureHttpMessageHandlerBuilder(this IHttpCl
...
             ReserveClient(builder, typeof(TClient), builder.Name, validateSingleType);
 
-            builder.Services.AddTransient<TClient>(s =>
+            builder.Services.AddTransient<TClient>(s => AddTransientHelper<TClient>(s));
+
+            return builder;
+
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2091:RequiresUnreferencedCode",
+                Justification = "Workaround for https://github.com/mono/linker/issues/1416. Outer method has been annotated with DynamicallyAccessedMembers.")]
+            TClient1 AddTransientHelper<TClient1>(IServiceProvider s)
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.cs:344]: Maybe `TClient1` => `TClientInner`?

Or better yet, maybe don't make the method nested.... Then you can just name it `TClient`.

File: src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.cs (hunk 1 above)
Comment (joperezr) [reply src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.cs:410]: Oh good catch, yeah I usually use that too, this was a copy-paste error 😩

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**